### PR TITLE
add "!" notation to remove optional groups with missing required children

### PR DESCRIPTION
**Pull request by rettigl**
If an optional group with a required child is set, but the data for the required child cannot be found, it is missing in the dict. This produces a warning during conversion.
This PR adds the option to add "!" as first element in a string value of the config file, to mark this field as required. If it is missing after filling the dict from all sources, all elements from this group are being removed from the template.

### CODE REVIEW

File: pynxtools_mpes/reader.py [hunk 1]
```diff
@@ -330,7 +330,13 @@ def read(  # pylint: disable=too-many-branches
 
         fill_data_indices_in_config(config_file_dict, x_array_loaded)
 
+        optional_groups_to_remove = []
+
         for key, value in config_file_dict.items():
+            if isinstance(value, str) and "!" in value and value.index("!") == 0:
```
Comment (domna) [inline pynxtools_mpes/reader.py:336]:
```suggestion
            if isinstance(value, str) and value.startswith("!"):
```

File: pynxtools_mpes/reader.py [hunk 2]
```diff
@@ -384,6 +390,18 @@ def read(  # pylint: disable=too-many-branches
         for key, value in eln_data_dict.items():
             template[key] = value
 
+        # remove groups that have required children missing
+        for key in optional_groups_to_remove:
+            if key not in template:
+                group_to_delete = "/".join(key.split("/")[:-1])
+                print(
```
Comment (domna) [inline pynxtools_mpes/reader.py:397]:
```suggestion
                logging.info(
```

File: pynxtools_mpes/reader.py [hunk 3]
```diff
@@ -384,6 +390,18 @@ def read(  # pylint: disable=too-many-branches
         for key, value in eln_data_dict.items():
             template[key] = value
 
+        # remove groups that have required children missing
+        for key in optional_groups_to_remove:
+            if key not in template:
+                group_to_delete = "/".join(key.split("/")[:-1])
```
Comment (domna) [inline pynxtools_mpes/reader.py:396]:
```suggestion
                group_to_delete = key.rsplit("/", 1)[0]
```

File: pynxtools_mpes/reader.py [hunk 4]
```diff
@@ -384,6 +390,18 @@ def read(  # pylint: disable=too-many-branches
         for key, value in eln_data_dict.items():
             template[key] = value
 
+        # remove groups that have required children missing
+        for key in optional_groups_to_remove:
+            if key not in template:
```
Comment (domna) [inline pynxtools_mpes/reader.py:395]:
```suggestion
            if key not in template or template[key] is None:
```

File: pynxtools_mpes/reader.py [hunk 5]
```diff
@@ -384,6 +390,18 @@ def read(  # pylint: disable=too-many-branches
...
+                group_to_delete = "/".join(key.split("/")[:-1])
+                print(
+                    f"[info]: Required element {key} not provided. "
+                    f"Removing the parent group {group_to_delete}.",
+                )
+                for temp_key in template.keys():
+                    if temp_key.find(group_to_delete) == 0:
```
Comment (domna) [inline pynxtools_mpes/reader.py:402]:
```suggestion
                    if temp_key.startswith(group_to_delete):
```

**Review by domna — APPROVED** [review]
I have a few suggestions for better understandability of the code otherwise it looks good.

**Comment by rettigl** [comment]
@domna Thanks, I was looking for the startswith, but somehow did not find it. Looks all fine to me, can you commit these changes, merge and make a new release? Thanks.

**Comment by rettigl** [comment]
@domna: I think your implementation of logging here does not do what it should, at least it does not print any messages for me. As far as I understand, that is because you defined no handler for the logger.
While this is handled similarly e.g. in helpers.py of pynxtools, there warnings are raised, which I think are printed by the default "last resort" handler.

**Pull request merged**